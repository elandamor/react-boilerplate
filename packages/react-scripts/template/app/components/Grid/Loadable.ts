/**
 *
 * Asynchronously loads the component for Grid
 *
 */

import React from 'react';

export default React.lazy(() => import('./index'));
