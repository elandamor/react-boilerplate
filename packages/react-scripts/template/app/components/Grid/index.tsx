import React, { FC } from 'react';
import classNames from 'classnames';
// Styles
import Wrapper from './styles';

export interface IProps {
  children: any;
  className?: string;
  columns: number;
  gap?: number;
}

/**
 * @render react
 * @name Grid component
 * @description Grid component.
 * @example
 * <Grid
 *  columns={2}
 *  gap={20}
 * >
 *  <Component />
 *  <Component />
 * </Grid>
 */

const Grid: FC<IProps> = ({ children, className, columns, gap }) => (
  <Wrapper
    className={classNames('c-grid', className)}
    columns={columns}
    gap={gap}
  >
    {children}
  </Wrapper>
);

export default Grid;
