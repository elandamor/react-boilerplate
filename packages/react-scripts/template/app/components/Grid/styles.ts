import styled from 'styled-components';
import { IGridProps } from './index';
import { THEME } from '../../global-styles';

const Wrapper = styled.div<IGridProps>`
  display: grid;
  grid-template-columns: repeat(${({ columns }) => columns}, 1fr);
  grid-gap: ${({ gap }) => (gap ? gap : THEME.space[3])}px;
`;

export default Wrapper;
