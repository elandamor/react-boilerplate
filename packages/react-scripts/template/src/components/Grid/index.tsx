import React, { FC } from 'react';
import classNames from 'classnames';
// Styles
import Wrapper from './styles';

export interface IGridProps {
  children: React.ReactNode;
  className?: string;
  columns: number;
  gap?: number;
}

/**
 * @render react
 * @name Grid component
 * @description Grid component.
 * @example
 * <Grid
 *  columns={2}
 *  gap={24}
 * >
 *  <Component />
 *  <Component />
 * </Grid>
 */

const Grid: FC<IGridProps> = ({ children, className, columns, gap }) => (
  <Wrapper
    className={classNames('c-grid', className)}
    columns={columns}
    gap={gap}
  >
    {children}
  </Wrapper>
);

export default Grid;
