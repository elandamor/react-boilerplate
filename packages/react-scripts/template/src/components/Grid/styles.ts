import styled from 'styled-components';
import { IGridProps } from './index';
import theme from '../../theme';

const Wrapper = styled.div<IGridProps>`
  display: grid;
  grid-template-columns: repeat(${({ columns }) => columns}, 1fr);
  grid-gap: ${({ gap }) => (gap ? gap : theme.space[3])}px;
`;

export default Wrapper;
