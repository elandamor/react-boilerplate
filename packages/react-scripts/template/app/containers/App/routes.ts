// tslint:disable:object-literal-sort-keys
import { Home } from '../../pages';

import { IRouteProps } from '../../components/Routes';

const routes: IRouteProps[] = [
  {
    exact: true,
    path: '/',
    component: Home,
  },
];

export default routes;
