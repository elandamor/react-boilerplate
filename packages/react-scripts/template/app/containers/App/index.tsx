import classNames from 'classnames';
import React, { Component, Suspense } from 'react';
import Measure from 'react-measure';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
// Components
import { ErrorBoundary, LoadingBar, Routes } from '../../components';
// Contexts
import NetworkStatusProvider from '../../contexts/networkStatus.context';
// Routes
import routes from '../../routes';
// Styles
import GlobalStyles from '../../global-styles';
import Wrapper from './styles';

import { breakpoints, makeDebugger } from '../../utils';
const debug = makeDebugger('App');

const defaultTheme = {
  breakpoints: {
    small: '600px',
    medium: '1024px',
    large: '1440px',
    xLarge: '1920px',
  },
};

/* tslint:disable:object-literal-sort-keys */
const themeLight = {
  ...defaultTheme,
  isDark: false,
  palette: {
    bodyBackground: '#FAFAFA',
    brandPrimary: '#FFFFFF',
    cardBackground: '#FFFFFF',
    cardBorderColor: '#E4E6E9',
  },
};
/* tslint:enable:object-literal-sort-keys */

export interface IProps extends RouteComponentProps<any> {}

interface IState {
  bounds: {
    height: number;
    width: number;
  };
  theme: object;
}

/**
 * @render react
 * @name App container
 * @description The skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar).
 */
class App extends Component<IProps, IState> {
  protected componentIsMounted: boolean;
  protected previousLocation: object;

  constructor(props: IProps) {
    super(props);

    this.state = {
      bounds: {
        height: 0,
        width: 0,
      },
      theme: themeLight,
    };

    this.previousLocation = props.location;
  }

  public componentDidMount() {
    this.componentIsMounted = true;
    debug('componentIsMounted');
  }

  public componentWillUpdate(nextProps: IProps) {
    const { location } = this.props;

    if (
      nextProps.history.action !== 'POP' &&
      (!location.state || !location.state.modal)
    ) {
      this.previousLocation = this.props.location;
    }
  }

  public componentWillUnmount() {
    this.componentIsMounted = false;
  }

  public setState(nextState: any, cb?: () => void) {
    if (this.componentIsMounted) {
      super.setState(nextState, cb);
    }
  }

  public render() {
    const { location } = this.props;
    const {
      bounds: { width },
    } = this.state;

    const isModal = !!(
      location.state &&
      location.state.modal &&
      this.previousLocation !== location
    );

    return (
      <ThemeProvider theme={this.state.theme}>
        <NetworkStatusProvider>
          <Measure
            bounds
            onResize={(contentRect) => {
              this.setState({ bounds: contentRect.bounds });
            }}
          >
            {({ measureRef }) => (
              <Wrapper
                className={classNames('c-app__container', breakpoints(width))}
                // @ts-ignore
                ref={measureRef}
              >
                <GlobalStyles />
                <ErrorBoundary>
                  <Suspense fallback={<LoadingBar loading />}>
                    <Routes
                      location={isModal ? this.previousLocation : location}
                      routes={routes}
                    />
                  </Suspense>
                </ErrorBoundary>
              </Wrapper>
            )}
          </Measure>
        </NetworkStatusProvider>
      </ThemeProvider>
    );
  }
}

export default withRouter(App);
