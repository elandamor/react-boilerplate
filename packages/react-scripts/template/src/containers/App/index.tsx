import React, { Suspense } from 'react';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { Normalize } from 'styled-normalize';
import get from 'lodash/get';
import merge from 'lodash/merge';
// Components
import { ErrorBoundary, Header, LoadingBar, Routes } from '@app/components';
// Contexts
import NetworkStatusProvider from '@app/contexts/networkStatus.context';
// Routes
import routes from '@app/routes';
// Styles
import { Wrapper } from './styles';

import GlobalStyles from '@app/global-styles';

import baseTheme from '@app/theme';
import { useDarkMode } from '@app/hooks';

// import { makeDebugger } from '@app/utils';
// const debug = makeDebugger('App');

export interface IAppProps extends RouteComponentProps {}

export const AppContext = React.createContext({});

/**
 * @render react
 * @name App container
 * @description The skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar).
 */

const App = () => {
  const [darkMode, setDarkMode] = useDarkMode();

  // Merge the color mode with the base theme to create a new theme object
  const getTheme = (mode: string) => merge({}, baseTheme, {
    colors: get(baseTheme.colors.modes, mode, baseTheme.colors),
    isDark: true,
  })

  const theme = darkMode ? getTheme('dark') : baseTheme;

  return (
    <ThemeProvider theme={theme}>
      <NetworkStatusProvider>
        <AppContext.Provider value={{ darkMode, setDarkMode }}>
          <Wrapper>
            <Normalize />
            <GlobalStyles />
            <ErrorBoundary>
              <Header alignItems="center" flex="none" />
            </ErrorBoundary>
            <ErrorBoundary>
              <Suspense fallback={<LoadingBar />}>
                <Routes routes={routes} />
              </Suspense>
            </ErrorBoundary>
          </Wrapper>
        </AppContext.Provider>
      </NetworkStatusProvider>
    </ThemeProvider>
  );
}

export default withRouter(App);
