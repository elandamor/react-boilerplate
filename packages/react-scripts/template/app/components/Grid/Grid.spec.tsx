// Grid.spec.tsx
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import 'jest-styled-components';
import 'whatwg-fetch';

import { ThemeProvider } from 'styled-components';

import Grid from './index';
import { themeLight } from '../../containers/App';

const theme = {
  ...themeLight,
};

describe('<Grid />', () => {
  it('should display a grid', () => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <Grid columns={2}>
          <div>1</div>
          <div>2</div>
          <div>3</div>
        </Grid>
      </ThemeProvider>
    ).toJSON();

    expect(component).toMatchSnapshot();
  });
});
