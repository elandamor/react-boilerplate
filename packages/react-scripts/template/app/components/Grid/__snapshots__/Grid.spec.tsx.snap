// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> should display a grid 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(2,1fr);
  grid-gap: 20px;
}

@media screen and (max-width:1024px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
  }
}

<div
  className="c-grid c0"
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;
